start_admin_sshd.sh

```diff
@@ -8,36 +8,87 @@ mkdir -p /home/ec2-user/.ssh/
-    continue
+set -f
+IFS='
+'
+
+# Is user-data does not contain a JSON block, there's not much point in continuing.
+if ! jq empty ${user_data} &>/dev/null; then
+    echo "${user_data} does not contain valid JSON." >&2
+fi
+
+# Populate authorized_keys with all the public keys found in api user-data
+if grep -q authorized_keys ${user_data}; then
+  ssh_authorized_keys="${ssh_config_dir}/authorized_keys"
+  touch ${ssh_authorized_keys}
+  chmod 600 ${ssh_authorized_keys}
+
+  userdata_authorized_keys=($(jq -r '.ssh.authorized_keys[]' ${user_data} 2> /dev/null))
+
+  for public_key_data in "${userdata_authorized_keys[@]}"; do
+    if [[ ! "${public_key_data}" =~ ^"ssh" ]]; then
+      echo "Key ${public_key_data} looks invalid" >&2
+      continue
+    fi
+    echo "${public_key_data}" >> "${ssh_authorized_keys}"
+    if ! grep -q "${public_key_data}" "${ssh_authorized_keys}"; then
+      echo "Failed to write key ${public_key_data} to ${ssh_authorized_keys}" >&2
+      continue
+    fi
+  done
+
+  # If we didn't write any keys at all, then this isn't a valid auth method
+  if [ ! -s "${ssh_authorized_keys}" ]; then
+    echo "Failed to write any valid public keys to ${ssh_authorized_keys}" >&2
+  else
+    ((++available_auth_methods))
   fi
-  echo "${public_key_data}" >> "${ssh_authorized_keys}"
-  if ! grep -q "${public_key_data}" "${ssh_authorized_keys}"; then
-    echo "Failed to write key with index ${public_key_index} to authorized_keys" >&2
-    continue
+fi
+
+# Populate trusted with all the trusted_user_ca_keys found in api user-data
+if grep -q trusted_user_ca_keys ${user_data}; then
+  ssh_trusted_user_ca_keys="${ssh_config_dir}/trusted"
+  touch ${ssh_trusted_user_ca_keys}
+  chmod 600 ${ssh_trusted_user_ca_keys}
+
+  userdata_trusted_user_ca_keys=($(jq -r '.ssh.trusted_user_ca_keys[]' ${user_data} 2> /dev/null))
+
+  for trusted_key_data in "${userdata_trusted_user_ca_keys[@]}"; do
+    if [[ ! "${trusted_key_data}" =~ ^"ssh" ]]; then
+      echo "Key ${trusted_key_data} looks invalid" >&2
+      continue
+    fi
+    echo "${trusted_key_data}" >> "${ssh_trusted_user_ca_keys}"
+    if ! grep -q "${trusted_key_data}" "${ssh_trusted_user_ca_keys}"; then
+      echo "Failed to write key ${trusted_key_data} to ${ssh_trusted_user_ca_keys}" >&2
+      continue
+    fi
+  done
+
+  # If we didn't write any keys at all, then this isn't a valid auth method
+  if [ ! -s "${ssh_trusted_user_ca_keys}" ]; then
+    echo "Failed to write any valid public keys to ${ssh_trusted_user_ca_keys}" >&2
+  else
+    printf "\nTrustedUserCAKeys: %s\n" "${ssh_trusted_user_ca_keys}" >> ${sshd_config}
+    # If we didn't append the TrustedUserCAKeys, then this isn't a valid auth method
+    if ! grep -q "^TrustedUserCAKeys: ${ssh_trusted_user_ca_keys}$" ${sshd_config}; then
+      echo "Failed to write TrustedUserCAKeys to sshd_config" >&2
+    else
+      ((++available_auth_methods))
```

Comment:
Typically `++n` is used in very specific use cases and tickles my spideysenses - can we switch to `n++`?

```suggestion
      ((available_auth_methods++))
```

Reply:
I gave this a shot, but the container seemed to stop immediately after the first instance:
```
+ [[ ! -s /home/ec2-user/.ssh/authorized_keys ]]
+ (( available_auth_methods++ ))
```
I think this is due to the `set -e` at the top of the script.

Reply:
Ah, yeah this would have to written as `: (( available_auth_methods++ ))` or `let available_auth_methods++`.

Reply:
It seems those also are causing an early exit.

Reply:
> It seems those also are causing an early exit.

TIL:

> from `bash -c 'help let'`
>
>     Exit Status:
>     If the last ARG evaluates to 0, let returns 1; let returns 0 otherwise.

The `: ((available_auth_methods++))` expression should be okay no matter what (because of `:` noop).

<details>

```bash
( x=0; set -x; : x++; echo $?; x=0; let x++; echo $?; x=0; let ++x; echo $?)
# + : x++
# + echo 0
# 0
# + x=0
# + let x++
# + echo 1
# 1
# + x=0
# + let ++x
# + echo 0
# 0
```

</details>